File: algorithm/cpp/prob501-1000/prob576_medium_Out-of-Boundary-Paths.cpp
// prob576: Out of Boundary Paths

/*
 * There is an m by n grid with a ball. Given the start coordinate (i,j) of the ball,
 * you can move the ball to adjacent cell or cross the grid boundary in four directions (up, down, left, right).
 * However, you can at most move N times. Find out the number of paths to move the ball out of grid boundary.
 * The answer may be very large, return it after mod 109 + 7.
 */

/*
 * Note:
 * Once you move the ball out of boundary, you cannot move it back.
 * The length and height of the grid is in range [1,50].
 * N is in range [0,50].
 */

#include <vector>

using namespace std;

class Solution {
public:
    int findPaths(int m, int n, int N, int i, int j) {
        if(N == 0) return 0;
        // i, j, k: m, n, N
        vector<vector<vector<int> > > dp(m, vector<vector<int> >(n, vector<int>(N + 1, -1)));
        return dfs(i, j, m, n, N, dp);
    }

private:
    int MOD = 1e9 + 7;

    int dfs(int i, int j, int m, int n, int k, vector<vector<vector<int> > >& dp)
    {
        if(i < 0 || i >= m || j < 0 || j >= n)
            return 1;
        if(k == 0)
            return 0;
        if(dp[i][j][k] != -1)
            return dp[i][j][k];

        dp[i][j][k] = 0;
        int dx[4] = {0, 1, 0, -1};
        int dy[4] = {1, 0, -1, 0};
        for(int d = 0; d < 4; ++d)
        {
            int x = i + dx[d];
            int y = j + dy[d];
            dp[i][j][k] = (dp[i][j][k] + dfs(x, y, m, n, k - 1, dp)) % MOD;
        }
        return dp[i][j][k];
    }
};

File: algorithm/python/prob501-1000/prob576_medium_Out-of-Boundary-Paths.py
import functools

MOD = int(1e9 + 7)

class Solution:
    def findPaths(self, m: int, n: int, maxMove: int, startRow: int, startColumn: int) -> int:
        self.m = m
        self.n = n
        self.dx = [0, 1, 0, -1]
        self.dy = [1, 0, -1, 0]
        if maxMove == 0:
            return 0
        return self.dfs(startRow, startColumn, maxMove)

    @functools.lru_cache(int(1e8))
    def dfs(self, i, j, k):
        if i < 0 or i >= self.m or j < 0 or j >= self.n:
            return 1
        if k == 0:
            return 0
        ans = 0
        for d in range(4):
            x = i + self.dx[d]
            y = j + self.dy[d]
            ans = (ans + self.dfs(x, y, k - 1)) % MOD
        return ans

